ext/EvoTreesCUDAExt/fit.jl

```diff
@@ -1,267 +1,337 @@
-        offset = 0 # identifies breakpoint for each node set within a depth
-        n_next = Int[]
-
-        # pred leafs if max depth is reached
-        if depth == params.max_depth
-            for n ∈ n_current
-                if L <: Quantile
-                    pred_leaf_cpu!(tree.pred, n, nodes[n].∑, L, params, ∇, nodes[n].is)
-                else
-                    pred_leaf_cpu!(tree.pred, n, nodes[n].∑, L, params)
-                end
-            end
-        else
-            # find best split for each node
-            for n ∈ n_current[1:2:end]
-                update_hist!(nodes[n].h, h∇, ∇, x_bin, nodes[n].is, jsg)
+    cache::EvoTrees.CacheGPU,
+    is::CuVector,
+) where {L,K}
+
+    backend = KernelAbstractions.get_backend(cache.x_bin)
+
+    ∇_gpu = cache.∇
+    if L <: EvoTrees.MAE
+        ∇_gpu = copy(cache.∇)
+        ∇_gpu[2, :] .= 1.0f0
+    end
+
+    # Initialize cache arrays
+    cache.tree_split_gpu .= false
+    cache.tree_cond_bin_gpu .= 0
+    cache.tree_feat_gpu .= 0
+    cache.tree_gain_gpu .= 0
+    cache.tree_pred_gpu .= 0
+    cache.nodes_sum_gpu .= 0
+    cache.anodes_gpu .= 0
+    cache.n_next_gpu .= 0
+    cache.n_next_active_gpu .= 0
+    cache.best_gain_gpu .= 0
+    cache.best_bin_gpu .= 0
+    cache.best_feat_gpu .= 0
+    cache.nidx .= 1
+    view(cache.anodes_gpu, 1:1) .= 1
+
+    # Root node processing
+    if params.max_depth == 1
+        reduce_root_sums_kernel!(backend)(
+            cache.nodes_sum_gpu, ∇_gpu, is;
+            ndrange=length(is),
+            workgroupsize=256,
+        )
+        KernelAbstractions.synchronize(backend)
+        n_active = 0
+    else
+        update_hist_gpu!(
+            cache.h∇, ∇_gpu, cache.x_bin, cache.nidx, cache.js, is,
+            1, view(cache.anodes_gpu, 1:1), cache.nodes_sum_gpu, params,
+            cache.feattypes_gpu, cache.monotone_constraints_gpu, cache.K,
+            view(cache.sums_temp_gpu, 1:(2*cache.K+1), 1:1),
+            cache.target_mask_buf, backend,
+        )
+
+        find_split_root! = find_best_split_from_hist_kernel!(backend)
+        find_split_root!(
+            L, view(cache.best_gain_gpu, 1:1),
+            view(cache.best_bin_gpu, 1:1),
+            view(cache.best_feat_gpu, 1:1),
+            cache.h∇, cache.nodes_sum_gpu,
+            view(cache.anodes_gpu, 1:1),
+            cache.js, cache.feattypes_gpu, cache.monotone_constraints_gpu,
+            params.lambda, params.L2, params.min_weight,
+            cache.K, view(cache.sums_temp_gpu, 1:(2*cache.K+1), 1:1);
+            ndrange=1,
+            workgroupsize=64,
+        )
+        KernelAbstractions.synchronize(backend)
+        n_active = 1
+    end
+
+    # Main loop: build tree level by level - FIX: stop at max_depth-1
+    for depth in 1:(params.max_depth-1)
+        iszero(n_active) && break
+
+        view(cache.n_next_active_gpu, 1:1) .= 0
+        n_nodes = 2^(depth - 1)
+        active_nodes = view(cache.anodes_gpu, 1:n_nodes)
```

Comment:
Why not just set `active_nodes = view(cache.anodes_gpu, 1:n_active)`?
AFAIU, the functions `find_split_all!` or `apply_splits_kernel!` should only need the list of active nodes (as no op should happen on inactive ones.